_posts/en/newsletters/2022-02-09-newsletter.md

```diff
@@ -0,0 +1,114 @@
+  developers replied with comments on Zhao's summary and her proposals,
+  including with additional or alternative proposals for changes that
+  could be made.
+
+    Discussion appeared to be ongoing as this summary was being written.
+
+## Bitcoin Core PR Review Club
+
+*In this monthly section, we summarize a recent [Bitcoin Core PR Review Club][]
+meeting, highlighting some of the important questions and answers.  Click on a
+question below to see a summary of the answer from the meeting.*
+
+FIXME:glozow
+
+{% include functions/details-list.md
+  q0="FIXME"
+  a0="FIXME"
+  a0link="https://bitcoincore.reviews/23443#l-29"
+%}
+
+## Releases and release candidates
+
+*New releases and release candidates for popular Bitcoin infrastructure
+projects.  Please consider upgrading to new releases or helping to test
+release candidates.*
+
+- [LND 0.14.2-beta.rc1][lnd 0.14.2-beta] is the release candidate for a
+  maintenance version that includes several bug fixes and a few minor
+  improvements.
+
+<!-- FIXME:harding to update Tuesday -->
+
+## Notable code and documentation changes
+
+*Notable changes this week in [Bitcoin Core][bitcoin core repo],
+[C-Lightning][c-lightning repo], [Eclair][eclair repo], [LDK][ldk repo],
+[LND][lnd repo], [libsecp256k1][libsecp256k1 repo], [Hardware Wallet
+Interface (HWI)][hwi repo], [Rust Bitcoin][rust bitcoin repo], [BTCPay
+Server][btcpay server repo], [BDK][bdk repo], [Bitcoin Improvement
+Proposals (BIPs)][bips repo], and [Lightning BOLTs][bolts repo].*
+
+- [Bitcoin Core #23508][] Add getdeploymentinfo RPC FIXME:Xekyo
+
+- [Bitcoin Core #21851][] release: support cross-compiling for arm64-apple-darwin FIXME:adamjonas
+
+- [Bitcoin Core #16795][] updates the `getrawtransaction`, `gettxout`,
+  `decoderawtransaction`, and `decodescript` RPCs to return the inferred
+  [output script descriptor][topic descriptors] for any scriptPubKeys
+  that are decoded.
+
+- [LND #6626][] sets 5% as the default fee for payments routed through
+  LN when created using the legacy `SendPayment`, `SendPaymentSync`, and
+  `QueryRoutes` RPCs.  Payments sent using the newer `SendPaymentV2` RPC
+  default to zero fees, essentially requiring users to specify a value.
+  An additional merged PR, [LND #6234][], defaults to 100% fees for
+  payments of less than 1,000 satoshis made with the legacy RPCs.
+
+- [LND #6177][] allows the users of the [HTLC][topic HTLC] interceptor
+  to specify the reason an HTLC was failed, making the interceptor more
+  useful for testing how failures affect software using LND.
+
+- [Rust-Lightning #1227][] improves the route-finding logic to account for known
+  historical payment failures/successes. These failures/successes are used to
+  determine the upper and lower bounds of channel balances, which gives the
+  route-finding logic a more accurate success probability when evaluating
+  routes. This is an implementation of the algorithm described in the paper by
```

Comment:
Maybe the following comments are too long and expressive but I had just spend quite some time studying the PR and and the interesting research applied in this PR and it seems to me that the current summary of PR 1227 is a bit inaccurate. I will try to lay out my understanding (please correct me if I should misunderstand something)

If I understand correctly Pickhardt and Richter suggest to address the question of splitting multipart payments optimally by using [minimum cost flows](https://en.wikipedia.org/wiki/Minimum-cost_flow_problem) instead of searching for shortest paths of smaller amounts independently after splitting the full amount. This was was indeed addressed in [news163](https://bitcoinops.org/en/newsletters/2021/08/25/) as indicated in the current draft of the newsletter.

However the PR 1227 of LDK seems not to add a minimum cost flow solver but focuses on adding Probabilistic channel scoring as a feature to the cost objective. This seems to be an idea that was previously introduced by Pickhardt et. al. and explained in [news142](https://bitcoinops.org/en/newsletters/2021/03/31/). Besides adding probabilistic channel scoring as an additional feature (besides routing fees) as discussed in [news142](https://bitcoinops.org/en/newsletters/2021/03/31/) (and as done by c-lighting and discussed in [news172](https://bitcoinops.org/en/newsletters/2021/10/27/)) the current PR 1227 also followed through the theoretical ideas explained in the [probabilistic path finding paper](https://arxiv.org/abs/2103.08576) to compute conditional probabilities if one previously observed successful or unsuccessful routing attempts. (I believe this is usually called [Bayesian learning or Bayesian interference](https://en.wikipedia.org/wiki/Bayesian_inference) even though - and to my surprise -  the authors of the two papers didn't use those terms)

Those ideas were also made more explicit in the work by Pickhardt and Richter by stating that nodes should maintain the so called `Uncertainty Network`. This is an additional view on the channel graph which encodes the success probabilities given the learnt knowledge about the liquidity in the channel from prior payment attempts with the help of evaluating conditional probabilities. This is indeed achieved by maintaining the upper and lower bounds of the `liquidity` in the channels. This was summarized with explicit formulas on how to update the probabilities to learn from previous failed and successful attempts by Pickhardt in the LDK issue tracker at https://github.com/lightningdevkit/rust-lightning/issues/1170#issuecomment-972396747 before PR 1227 was opened

The question of multi part splitting seems not to be addressed or impoved by LDK yet as indicated in issue https://github.com/lightningdevkit/rust-lightning/issues/1276.

Reply:
Ah my bad I linked the wrong paper! 🤦 

I think my general description is still somewhat accurate, is there an important detail that you think I've missed that our readers would like summarized?

Reply:
I think it's no surprise you confused the two paper. The author of PR 1227 also links to the second paper on minimum cost flows - where (to be fair) the uncertainty network was explicitly introduced by applying the section of information gain and conditional probabilities from the first paper on probabilistic path finding. 

With respect to your question: I think it might make sense for the readers of the newsletter to recommend the summary of the two papers (and mention both?) which is given in https://github.com/lightningdevkit/rust-lightning/issues/1170#issuecomment-972396747. I think that this text and the following discussion extracts the theoretical knowledge from both papers to a very practical hands on guide for engineers and developers. It certainly has helped me to understand what is going on there and how to 

1. maintain the Uncertainty Network by keeping the minimum and maximum liquidity (as stressed in your text) and learning those boundaries properly from failed and successful attempts. 
2.  Translate the liquidity boundaries to conditional probabilities

The later seems to be (besides adding probabilistic channel scoring to the cost function in the first step) the core contribution of PR 1227. Adding the conditional probabilities also goes far beyond the c-lightning implementation of probabilistic scoring - which IIRC according to the evaluation of Christian Decker has demonstrated substantial improvements by decreasing failure rate, overall time to complete and number of attempts by roughly 50%

Reply:
I think it's no surprise that you confused the two paper. The author of PR 1227 also links to the second paper on minimum cost flows - where (to be fair) the uncertainty network was explicitly introduced by applying the section of information gain and conditional probabilities from the first paper on probabilistic path finding. 

With respect to your question: I think it might make sense for the readers of the newsletter to recommend the summary of the two papers (and mention both?) which is given in https://github.com/lightningdevkit/rust-lightning/issues/1170#issuecomment-972396747. I think that this text and the following discussion extracts the theoretical knowledge from both papers to a very practical hands on guide for engineers and developers. It certainly has helped me to understand what is going on there and how to 

1. maintain the Uncertainty Network by keeping the minimum and maximum liquidity (as stressed in your text) and learning those boundaries properly from failed and successful attempts. 
2.  Translate the liquidity boundaries to conditional probabilities

The later seems to be (besides adding probabilistic channel scoring to the cost function in the first step) the core contribution of PR 1227. Adding the conditional probabilities also goes far beyond the c-lightning implementation of probabilistic scoring. The previous c-lightning change has - [according to the blog post by Christian Decker](https://medium.com/blockstream/c-lightning-v0-10-2-bitcoin-dust-consensus-rule-33e777d58657) - demonstrated substantial improvements by decreasing failure rate, overall time to complete and number of attempts by roughly 30 to 50%. Unfortunately there was no evaluation or benchmark conducted for the additional improvements that are introduced by PR 1227.

Other than that (but probably too detailed for the newsletter): 
1. PR 1227 introduces a half life of 3600 seconds to forget the learnt boundaries and this choice was somewhat arbitrary.
2. Similar to the c-lightning implementation there was quite some discussion on how to combine the probabilistic channel scoring with the fee based scoring (weighted arithmetic mean vs harmonic or geometric mean). It seems no satisfying solution was found.